book/src/java_package.md

```diff
@@ -1,4 +1,180 @@
 # The `java_package` macro
 
-The `java_package` macro creates Rust structures to interact with a java package. It uses `javap` to read the class files and find the existing methods and other details so that you don't have to do any manual work.
+The `java_package` macro creates Rust structures to interact with a java package -- we call this *oxidizing* the Java class. Here is an example input to `java_package` that shows off the various features.
 
+```java
+package my.package;
+
+// Oxidize a class, ignoring all its methods or other details. This is useful
+// for classes that the Rust code needs to pass around opaquely but doesn't have to
+// actually use.
+class SimpleClass { }
+
+// Oxidize a class with all details inferred via Java reflection. This will cause
+// compilation errors if the class employs Java features that can't be supported
+// by duchess in Rust, such as overloaded functions or some of the richer uses
+// of Java wildcards (e.g., `ArrayList<Class<?>>`)
+//
+// Careful: since Java's semver rules are different from Rust's rules,
+// this can cause breakage if you update the Java package without updating
+// to a new Rust major version. For example, the java package might add a new
+// overloaded function; this is not a breaking change in Java, but it will cause
+// an error in your Rust crate.
+//
+// Therefore, we recommend that libraries which wish to maintain a semver guarantee
+// avoid this form.
+class ReflectedClass { * }
+
+// The preferred form is to specify exactly which parts of the Java API you wish
+// to include in the oxidized Rust type. This format is precisely the same as the
+// one generated by `javap -public`, so we recommend that you simply run that tool
+// and copy-and-paste the output in. You can then remove any methods or other details that
+// are causing trouble.
+class SpecifiedClass 
+    extends some.SuperClass           // Must be some superclass of `SpecifiedClass`
+    implements some.Interface1,
+               some.Interface2<Type>  // Must be interfaces implemented by `SpecifiedClass` or some superclass
+{
+    // Mirror a constructor by using the name of the class, along with types
+    // for its arguments. Note that we use full types. You can generate these signatures
+    // with `javap -public`.
+    SpecifiedClass(java.lang.String, java.util.List<String>);
+
+    // Mirror a method with the method name and argument types:
+    methodName(byte[], int);
+}
+```
+
+## Generated Rust code
+
+This will generate a Rust module structure containing:
+
+* a module for each Java package
+* for each oxidized Java class `Foo`:
+    * a struct `Foo` and a trait `FooExt` for each oxidized Java class `Foo`
+        * the trait defines methods on `Foo` that can be invoked on any [JVM operation](./jvm_operations.md) that returns a `Foo`.
+    * impls of the `JRef` trait for each superclass and interface, to permit upcasting
```

Comment:
I presume this is an upcoming refactor to rename the upcast trait?

Reply:
Oh yes I didn't get to that. I forget why but I had convinced myself it would be better to use a special trait vs `AsRef`.

Reply:
IIRC there was some impl conflict between some of our types when we tried to model upcasts through just `AsRef`. 

Also I think we need to be careful about who can implement the upcast. I think `call_method_unchecked` can blow up the JVM if a user made an erroneous `JRef` impl and invoked a JNI method that doesn't actually existing for some type.